lib/internal/Magento/Framework/Controller/Result/Json.php

```diff
@@ -29,24 +29,63 @@ class Json extends AbstractResult
     protected $json;
 
     /**
-     * @param \Magento\Framework\Translate\InlineInterface $translateInline
+     * @var \Magento\Framework\Serialize\Serializer\Json
      */
-    public function __construct(InlineInterface $translateInline)
-    {
+    private $serializer;
+
+    /**
+     * @param InlineInterface $translateInline
+     * @param \Magento\Framework\Serialize\Serializer\Json|null $serializer
+     * @throws \RuntimeException
+     */
+    public function __construct(
+        InlineInterface $translateInline,
+        \Magento\Framework\Serialize\Serializer\Json $serializer = null
+    ) {
         $this->translateInline = $translateInline;
+        $this->serializer = $serializer ?: \Magento\Framework\App\ObjectManager::getInstance()
+            ->get(\Magento\Framework\Serialize\Serializer\Json::class);
     }
 
     /**
      * Set json data
      *
-     * @param mixed $data
-     * @param boolean $cycleCheck Optional; whether or not to check for object recursion; off by default
-     * @param array $options Additional options used during encoding
-     * @return $this
+     * @param array|string|\Magento\Framework\DataObject $data
+     * @param bool $cycleCheck
+     * @param array $options
+     * @return Json
+     * @throws \InvalidArgumentException
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     * @deprecated
      */
     public function setData($data, $cycleCheck = false, $options = [])
     {
-        $this->json = \Zend_Json::encode($data, $cycleCheck, $options);
+        if ($data instanceof \Magento\Framework\DataObject) {
+            return $this->setArrayData($data->toArray());
+        }
+
+        if (is_array($data)) {
+            return $this->setArrayData($data);
+        }
+
+        if (is_string($data)) {
+            return $this->setJsonData($data);
+        }
+
+        throw new \Magento\Framework\Exception\LocalizedException(
+            new \Magento\Framework\Phrase('Invalid argument type')
+        );
+    }
+
+    /**
+     * @param array $data
+     * @return $this
+     * @throws \InvalidArgumentException
+     */
+    public function setArrayData(array $data)
```

Comment:
This is what I strongly disagree with. Its responsibility is quite clear - encode value and set it. There is no need to think which value are're passing.

`setData` - pass value and it will be encoded
`setJsonData` - pass string and it will be set as is

I do agree that `setData($object)` could be discouraged and corresponding `if` could be marked as deprecated for further removal and `setData` method simplification. But introduction of `setArrayData` method looks like a mess to me.